google/resource_cloudfunctions_function.go

```diff
@@ -0,0 +1,468 @@
+	if err != nil {
+		return err
+	}
+
+	d.Set("name", cloudFuncId.Name)
+	d.Set("description", getOpt.Description)
+	d.Set("entry_point", getOpt.EntryPoint)
+	d.Set("available_memory_mb", getOpt.AvailableMemoryMb)
+	sRemoved := strings.Replace(getOpt.Timeout, "s", "", -1)
+	timeout, err := strconv.Atoi(sRemoved)
+	if err != nil {
+		return err
+	}
+	d.Set("timeout", timeout)
+	d.Set("labels", getOpt.Labels)
+	if getOpt.SourceArchiveUrl != "" {
+		sourceArr := strings.Split(getOpt.SourceArchiveUrl, "/")
+		d.Set("storage_bucket", sourceArr[2])
+		d.Set("storage_object", sourceArr[3])
+	}
+
+	if getOpt.HttpsTrigger != nil {
+		d.Set("trigger_http", true)
+	}
+	if getOpt.EventTrigger != nil {
+		switch getOpt.EventTrigger.EventType {
+		//From https://github.com/google/google-api-go-client/blob/master/cloudfunctions/v1/cloudfunctions-gen.go#L335
+		case "providers/cloud.pubsub/eventTypes/topic.publish":
+			d.Set("trigger_topic", extractLastResourceFromUri(getOpt.EventTrigger.Resource))
+		case "providers/cloud.storage/eventTypes/object.change":
+			d.Set("trigger_bucket", extractLastResourceFromUri(getOpt.EventTrigger.Resource))
+		}
+	}
+	d.Set("region", cloudFuncId.Region)
+	d.Set("project", cloudFuncId.Project)
+
+	return nil
+}
+
+func resourceCloudFunctionsUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG]: Updating google_cloudfunctions_function")
+	config := meta.(*Config)
+
+	cloudFuncId, err := parseCloudFunctionId(d.Id(), config)
+	if err != nil {
+		return err
+	}
+
+	d.Partial(true)
+
+	function := cloudfunctions.CloudFunction{
+		Name: cloudFuncId.cloudFunctionId(),
+	}
+
+	_, err = config.clientCloudFunctions.Projects.Locations.Functions.Get(function.Name).Do()
+
+	if err != nil {
+		return fmt.Errorf("Function %s doesn't exists.", d.Get("name").(string))
+	}
+
+	if d.HasChange("labels") {
+		function.Labels = expandLabels(d)
+
+		op, err := config.clientCloudFunctions.Projects.Locations.Functions.Patch(function.Name, &function).
+			UpdateMask("labels").Do()
+
+		if err != nil {
+			return fmt.Errorf("Error when updating labels: %s", err)
+		}
+
+		d.SetPartial("labels")
+		err = cloudFunctionsOperationWait(config.clientCloudFunctions, op,
+			"Updating CloudFunctions Function Labels")
+		if err != nil {
+			return err
+		}
+	}
+	configUpdate := false
+	var updateMaskArr []string
+	if d.HasChange("available_memory_mb") {
+		availableMemoryMb := d.Get("available_memory_mb").(int)
+		function.AvailableMemoryMb = int64(availableMemoryMb)
+		updateMaskArr = append(updateMaskArr, "availableMemoryMb")
+		configUpdate = true
+	}
+
+	if d.HasChange("description") {
+		function.Description = d.Get("description").(string)
+		updateMaskArr = append(updateMaskArr, "description")
+		configUpdate = true
+	}
+
+	if d.HasChange("timeout") {
+		function.Timeout = fmt.Sprintf("%vs", d.Get("timeout").(int))
+		updateMaskArr = append(updateMaskArr, "timeout")
+		configUpdate = true
+	}
+
+	if configUpdate {
```

Comment:
nit: instead of a boolean, you can just check len(updateMaskArr) > 0

Reply:
Fixed :)